x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/event_log/get_execution_event_aggregation/index.ts

```diff
@@ -0,0 +1,351 @@
+const SEARCH_DURATION_FIELD = 'kibana.alert.rule.execution.metrics.total_search_duration_ms';
+const STATUS_FIELD = 'kibana.alert.rule.execution.status';
+
+const ONE_MILLISECOND_AS_NANOSECONDS = 1_000_000;
+
+const SORT_FIELD_TO_AGG_MAPPING: Record<string, string> = {
+  timestamp: 'ruleExecution>executeStartTime',
+  duration_ms: 'ruleExecution>executionDuration',
+  indexing_duration_ms: 'securityMetrics>indexDuration',
+  search_duration_ms: 'securityMetrics>searchDuration',
+  gap_duration_ms: 'securityMetrics>gapDuration',
+  schedule_delay_ms: 'ruleExecution>scheduleDelay',
+  num_triggered_actions: 'ruleExecution>numTriggeredActions',
+  // TODO: To be added in https://github.com/elastic/kibana/pull/126210
+  // total_alerts_created: 'securityMetrics>totalAlertsDetected',
+  // total_alerts_detected: 'securityMetrics>totalAlertsCreated',
+};
+
+/**
+ * Returns `aggs` to be supplied to aggregateEventsBySavedObjectIds
+ * @param maxExecutions upper bounds of execution events to return (to narrow below max terms agg limit)
+ * @param page current page to retrieve, starting at 0
+ * @param perPage number of execution events to display per page
+ * @param sort field to sort on
+ */
+export const getExecutionEventAggregation = ({
+  maxExecutions,
+  page,
+  perPage,
+  sort,
+}: ExecutionEventAggregationOptions): Record<string, estypes.AggregationsAggregationContainer> => {
+  // Last stop validation for any other consumers so there's a friendly message instead of failed ES Query
+  if (maxExecutions > MAX_EXECUTION_EVENTS_DISPLAYED) {
+    throw new BadRequestError(
+      `Invalid maxExecutions requested "${maxExecutions}" - must be less than ${MAX_EXECUTION_EVENTS_DISPLAYED}`
+    );
+  }
+
+  if (page < 0) {
+    throw new BadRequestError(`Invalid page field "${page}" - must be greater than 0`);
+  }
+
+  if (perPage <= 0) {
+    throw new BadRequestError(`Invalid perPage field "${perPage}" - must be greater than 0`);
+  }
+
+  const sortFields = flatMap(sort as estypes.SortCombinations[], (s) => Object.keys(s));
+  for (const field of sortFields) {
+    if (!Object.keys(SORT_FIELD_TO_AGG_MAPPING).includes(field)) {
+      throw new BadRequestError(
+        `Invalid sort field "${field}" - must be one of [${Object.keys(
+          SORT_FIELD_TO_AGG_MAPPING
+        ).join(',')}]`
+      );
+    }
+  }
+
+  return {
+    // Total unique executions for given root filters
+    totalExecutions: {
+      cardinality: {
+        field: EXECUTION_UUID_FIELD,
+      },
+    },
+    executionUuid: {
+      // Bucket by execution UUID
+      terms: {
+        field: EXECUTION_UUID_FIELD,
+        size: maxExecutions,
+        order: formatSortForTermsSort(sort),
+      },
+      aggs: {
+        // Bucket sort for paging
+        executionUuidSorted: {
+          bucket_sort: {
+            sort: formatSortForBucketSort(sort),
+            from: page * perPage,
+            size: perPage,
+            gap_policy: 'insert_zeros',
```

Comment:
Comment on ES issue that we're using this: https://github.com/elastic/elasticsearch/issues/36487

Reply:
Added in d205841